src/main/java/org/broadinstitute/hellbender/tools/walkers/fasta/ShiftFasta.java

```diff
@@ -0,0 +1,210 @@
+        final Path path = IOUtils.getPath(output);
+        chainId = 1;
+        try {
+            refWriter = new FastaReferenceWriterBuilder()
+                    .setFastaFile(path)
+                    .setBasesPerLine(basesPerLine)
+                    .build();
+            chainFileWriter = new FileWriter(shiftBackOutput);
+            if (intervalFilename != null) {
+                intervalRegularWriter = new FileWriter(intervalFilename+ ".intervals");
+                intervalShiftedWriter = new FileWriter(intervalFilename + ".shifted.intervals");
+            }
+        } catch (IOException e) {
+            throw new UserException.CouldNotCreateOutputFile("Couldn't create " + output + ", encountered exception: " + e.getMessage(), e);
+        }
+    }
+
+    public void traverse() {
+        SAMSequenceDictionary refDict = refSource.getSequenceDictionary();
+        long refLengthLong = refDict.getReferenceLength();
+        if (refLengthLong > Integer.MAX_VALUE) {
+            // TODO fix this??
+            throw new UserException.BadInput("Reference length is too long");
+        }
+        List<SAMSequenceRecord> contigs = refSource.getSequenceDictionary().getSequences();
+        final ListIterator<Integer> shiftOffsetsIt = (shiftOffsets != null && shiftOffsets.size() == contigs.size()) ?
+                shiftOffsets.listIterator() : null;
+        refSource.getSequenceDictionary().getSequences().forEach(seq -> shiftContig(seq, shiftOffsetsIt));
+    }
+
+    protected void shiftContig(SAMSequenceRecord seq, ListIterator<Integer> shiftOffsetsIt) {
+        int contigLength = seq.getSequenceLength();
+        int shiftOffset = shiftOffsetsIt == null ? contigLength/2 : shiftOffsetsIt.next();
+        if (shiftOffset != 0) {
+            byte[] bases = refSource.queryAndPrefetch(new SimpleInterval(seq.getSequenceName(), 1, contigLength)).getBases();
+            byte[] basesAtEnd = Arrays.copyOfRange(bases, shiftOffset, bases.length);
+            byte[] basesAtStart = Arrays.copyOf(bases, shiftOffset);
+            int shiftBackOffset = bases.length - shiftOffset;
+
+            try {
+                refWriter.startSequence(seq.getSequenceName(), basesPerLine);
+                refWriter.appendBases(basesAtEnd).appendBases(basesAtStart);
+                chainFileWriter.append(createChainString(seq.getSequenceName(), shiftBackOffset, contigLength, shiftOffset, bases.length, 0, shiftBackOffset, chainId++));
+                chainFileWriter.append("\n" + shiftBackOffset + "\n\n");
+                chainFileWriter.append(createChainString(seq.getSequenceName(), shiftOffset - 1, contigLength, 0, shiftOffset, shiftBackOffset, bases.length, chainId++));
+                chainFileWriter.append("\n" + shiftOffset + "\n\n");
+                if (intervalFilename != null && shiftOffsetsIt == null) {
+                    int intervalStart = shiftOffset/2;
```

Comment:
I would compartmentalize these writers into their own methods to make it a bit clearer and safer.